jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -867,20 +867,27 @@ public long getLength()
         @Override
         public Content.Chunk read()
         {
-            HttpStream stream;
-            try (AutoLock ignored = _lock.lock())
+            Content.Chunk chunk;
+            try
             {
-                HttpChannelState httpChannel = lockedGetHttpChannelState();
+                HttpStream stream;
+                try (AutoLock ignored = _lock.lock())
+                {
+                    HttpChannelState httpChannel = lockedGetHttpChannelState();
 
-                Content.Chunk error = httpChannel._readFailure;
-                httpChannel._readFailure = Content.Chunk.next(error);
-                if (error != null)
-                    return error;
+                    Content.Chunk error = httpChannel._readFailure;
+                    httpChannel._readFailure = Content.Chunk.next(error);
+                    if (error != null)
+                        return error;
 
-                stream = httpChannel._stream;
+                    stream = httpChannel._stream;
+                }
+                chunk = stream.read();
+            }
+            catch (Throwable t)
+            {
+                return Content.Chunk.from(t, true);
```

Comment:
Shouldn't this notify onError() too?

Reply:
No. There is no `onError` in our `Content.Source` contract. There are various `onError` methods in layers above, and these should be called when a `Chunk.Error` is read.